modules/visual_script/editor/visual_script_editor.cpp

```diff
@@ -4342,6 +4360,44 @@ void VisualScriptEditor::update_toggle_scripts_button() {
 	toggle_scripts_button->set_tooltip(vformat("%s (%s)", TTR("Toggle Scripts Panel"), ED_GET_SHORTCUT("script_editor/toggle_scripts_panel")->get_as_text()));
 }
 
+void VisualScriptEditor::_script_class_icon_path_dialog_confirmed() {
+	const String &path = _script_class_icon_path_dialog->get_current_path();
+	_script_class_icon_btn_set_icon(path);
+}
+
+void VisualScriptEditor::_script_class_icon_path_dialog_file_selected(const String &p_path) {
+	_script_class_icon_btn_set_icon(p_path);
+}
+
+void VisualScriptEditor::_script_class_icon_btn_set_icon(const String &p_path) {
+	ERR_FAIL_COND(!_script_class_icon_btn);
+	ERR_FAIL_COND_MSG(p_path.length() && !FileAccess::exists(p_path), "The given script class icon path does not exist.");
+	RES icon = FileAccess::exists(p_path) ? ResourceLoader::load(p_path, "Texture") : (RES)EditorNode::get_singleton()->get_gui_base()->get_theme_icon("Nil", "EditorIcons");
+	undo_redo->create_action("Set script class icon");
+	undo_redo->add_do_method(_script_class_icon_btn, "set_button_icon", icon);
+	undo_redo->add_undo_method(_script_class_icon_btn, "set_button_icon", _script_class_icon_btn->get_icon());
+	if (script.is_valid()) {
+		undo_redo->add_do_method(script.operator->(), "set_script_class_icon_path", p_path);
+		undo_redo->add_undo_method(script.operator->(), "set_script_class_icon_path", script->get_script_class_icon_path());
+	}
+	undo_redo->commit_action();
+	set_edited(true);
+}
+
+void VisualScriptEditor::_script_class_name_text_changed(const String &p_text) {
+	if (script.is_valid()) {
+		script->set_script_class_name(p_text);
+	}
+	set_edited(true);
+}
```

Comment:
Actually, the script is checked for validity, but you're right anyway. If the script isn't valid, set_edited shouldn't be set to true either.

Same in the next line method for the icon path.

Reply:
Yes, that's right. Good fix.

Reply:
@dploeger I think you may have missed what Techno meant here (although you're right too). Techno is saying that you need to check that the `p_text` is a valid identifier, and NOT set the script class name if it fails that check. This then removes the need for performing the `is_valid_identifier()` check in the `_get_global_class_name` method for `VisualScriptLanguage` since assigning the script class name here already did the check for us.

Reply:
So what you saying is that the if clause in 4388 should be extended to check p_text.is_valid_identifier() as well (although is_valid_identifier is probably not a String method, right?)